Mathlib/Algebra/Group/Defs.lean

```diff
@@ -620,11 +620,139 @@ the file `Algebra.GroupPower.Basic`. For now, we only register the most basic pr
 need right away.
 -/
 
+/--
+An abbreviation for `npowRec` with the typeclass hypotheses matching `npowBinRec`,
+so that we can use `@[csimp]` to replace it with an implementation by repeated squaring in compiled
+code.
+-/
+@[to_additive (attr := nolint unusedArguments)
+"An abbreviation for `nsmulRec` with the typeclass hypotheses matching `nsmulBinRec`,
+so that we can use `@[csimp]` to replace it with an implementation by repeated doubling in compiled
+code."]
+abbrev npowRec' {M : Type*} [Semigroup M] [One M]
+    (_mul_one : ∀ m : M, m * 1 = m) (_one_mul : ∀ m : M, 1 * m = m) (k : ℕ) (m : M) : M :=
+  npowRec k m
+
+@[to_additive]
+theorem npowRec'_two_mul {M : Type*} [Semigroup M] [One M]
+    (mul_one : ∀ m : M, m * 1 = m) (one_mul : ∀ m : M, 1 * m = m) (k : ℕ) (m : M) :
+    npowRec' mul_one one_mul (2 * k) m = npowRec' mul_one one_mul k (m * m) := by
+  dsimp [npowRec']
+  induction k with
+  | zero => rfl
+  | succ n ih => simp [npowRec, ih, mul_assoc]
+
+/--
+Scalar multiplication by repeated self-addition,
+the additive version of exponentation by repeated squaring.
+-/
+-- Ideally this would be generated by `@[to_additive]` from `npowBinRec`.
+-- The unused arguments are essential so that it is possible to prove the `csimp` theorem
+-- `nsmulRec_eq_nsmulBinRec`
+@[nolint unusedArguments]
+def nsmulBinRec {M : Type*} [AddSemigroup M] [Zero M]
+    (_add_zero : ∀ m : M, m + 0 = m) (_zero_add : ∀ m : M, 0 + m = m) (k : ℕ) (m : M) : M :=
+  go k 0 m
+where
+  /-- Auxiliary tail-recursive implementation for `nsmulBinRec`-/
+  go : ℕ → M → M → M
+  | 0, y, _ => y
+  | (k + 1), y, x =>
+    let k' := (k + 1) >>> 1
+    if k &&& 1 = 1 then
+      go k' y (x + x)
+    else
+      go k' (y + x) (x + x)
```

Comment:
Here's a way to eliminate this repetition between the `smul` and `pow` versions:
```
/-- Fold `f` across the bits of `n` -/
def foldBits {σ} (f : Bool → σ → σ) (init : σ) (n : ℕ) : σ :=
  match n, init with
  | 0, s => s
  | (k + 1), s =>
    let k' := (k + 1) >>> 1
    foldBits f (f (k &&& 1 == 1) s) k'
  termination_by foldBits f _ k => k
  -- We use this `decreasing_by` block rather than inline `have` statements to avoid
  -- https://github.com/leanprover/std4/issues/428
  decreasing_by
    simp_wf
    exact Nat.div_lt_self (Nat.zero_lt_succ _) (Nat.lt_succ_self _)

def nsmulBinRec' {M : Type*} [AddSemigroup M] [Zero M]
    (_add_zero : ∀ m : M, m + 0 = m) (_zero_add : ∀ m : M, 0 + m = m) (k : ℕ) (m : M) : M :=
  Prod.fst <| foldBits (fun b p => (if b then p.1 else p.1 + p.2, p.2 + p.2)) (0, m) k
```

Reply:
Does this still produce something tail-recursive?

Reply:
It must be, right? `foldBits` is the head symbol of the recursion.

Reply:
I don't think I've ever written a tail recursive function via well founded recursion, and am not certain it actually works!

Reply:
@semorrison, I have no idea what you're talking about. The lines that I am suggesting a replacement for are a tail recursive function via well founded recursion, that I assume was written by you! All I've done is factor out the common piece.

Reply:
Eric's `foldBits` suggestion seems like it would generate fine code, so long as it's marked `@[inline]`. If you go with it, be sure to specify what "fold over the bits" means. Presumably from least- to most- significant bits? It seems to go until the zero bit right above the the most-significant set bit?

I'm not sure whether the compiler will split the product type into separate arguments, so it might actually produce slightly worse code. I don't think the code duplication here is that bad, and maybe if you're interested in an API for folding over bits @eric-wieser you can pursue this after this PR is merged?

Reply:
I think it's also worth considering (or ruling out) the approach in #9154 here, which claims that `binaryRec` works fine after all.

Reply:
> @semorrison, I have no idea what you're talking about. The original lines that I am suggesting a replacement for are themselves a tail recursive function via well founded recursion, that I assume was written by you! All I've done is factor out the common piece.

@eric-wieser, my function uses structural recursion on `Nat`, not well-founded recursion. These behave very differently.

Reply:
I've asked @negiizhao if they can update #9154; hopefully that will work out.